Enhance Load Generator to Query Pre-Generated Data Blocks

kushalShukla-web opened the pull request:
changed Load generator to query from downloaded blocks. Here i have added "current" and "absolute" time to query data between current data (head_blocks) and absolute data (Downloaded) blocks. 
Things left - i have to configure to add  instant query on downloaded blocks !

kushalShukla-web committed:
changed query in load-generator such that it will query from downloaded blocks.
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

kushalShukla-web committed:
Modified the query way ! added absolute and current time pluse changed loadgen
start and end time!
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

kushalShukla-web commented:
Instant query is still missing; currently, it does not perform instant queries on downloaded blocks.

kakkoyun commented on tools/load-generator/main.go:
```diff
@@ -139,27 +174,50 @@ func (q *Querier) run(wg *sync.WaitGroup) {
+		defer resp.Body.Close()
+
+		// Print the body content as a string
+		fmt.Println(string(body), "Hi from body side!!!")
```
It needs to a bit of clean up.

kushalShukla-web committed:
Added Instant query to downloaded blocks.
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

kushalShukla-web committed:
Removed Randomness from the code.
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

kushalShukla-web committed:
1. Replaced "key" Keyword  with "bucket" Keyword.
2. Replaced images to prominfra docker registery.
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

bboreham commented on tools/load-generator/main.go:
```diff
@@ -120,16 +132,59 @@ func NewQuerier(groupID int, target, prNumber string, qg QueryGroup) *Querier {
+	return &bucketConfig, nil
+}
+
+func configstate(v *BucketConfig, err error) *configState {
```
It is potentially confusing to have two identifiers which differ only in the case of some letters.

bboreham commented on tools/load-generator/main.go:
```diff
@@ -120,16 +132,59 @@ func NewQuerier(groupID int, target, prNumber string, qg QueryGroup) *Querier {
 	for {
 		start := time.Now()
-
+		runBlockMode := "current"
```
I think the idea here is to alternate between current and absolute?
At least, put a comment saying that.

bboreham commented on tools/load-generator/main.go:
```diff
@@ -120,16 +132,59 @@ func NewQuerier(groupID int, target, prNumber string, qg QueryGroup) *Querier {
+	}
+	return &configState{
+		bucketConfig: v,
+		Err:          err,
```
I think the idea is we carry `err` around, and act like this config doesn't exist if `err` is non-nil.

This at least deserves a comment.  Maybe an alternative would work, just having `nil` for this `configState` object, if there would be an error, and checking that?

bboreham commented on tools/load-generator/main.go:
```diff
@@ -120,16 +132,59 @@ func NewQuerier(groupID int, target, prNumber string, qg QueryGroup) *Querier {
-func (q *Querier) run(wg *sync.WaitGroup) {
+// Function to load `minTime` and `maxTime` from bucket-config.yml
+func loadKeyConfig() (*BucketConfig, error) {
+	filePath := "/config/bucket-config.yml"
```
This should probably be a parameter, a CLI flag.

kushalShukla-web committed:
1. Added comment section of absolute and current block.
2. Removed Err and absoluteTime from the configState object.
3. Added flag package for debugging purpose and removed hard coded file path.

Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

kakkoyun committed:
Merge branch 'master' into query-data

kushalShukla-web commented:
same like  PR https://github.com/prometheus/test-infra/pull/779 .
